Change `Tests.prototype.all_done` to require either tests or pending remotes

andreubotella opened the pull request:
`Tests.prototype.all_done` used to treat the presence of at least one test as a necessary condition for all tests to be done. This would be false in cases where an exception is thrown inside the `setup` function, but the code path that runs in that case doesn't use the `all_done()` function.

However, if an exception is thrown inside the `setup` function in a worker test, `tests.all_done()` is called in that code path, which results in the completion message from the worker being ignored. This change fixes this by changing the condition to require the presence of either tests or remote contexts.

~For service worker tests, however, that is not enough, since service worker registration is asynchronous and `tests.all_done()` will be called before the SW remote context is set up. This change lets the `fetch_tests_from_worker` function take a promise or an async function that returns the worker, and it guarantees that the test won't be considered complete until the promise is resolved. The HTML boilerplate for service worker tests is also changed in turn.~

Closes #29777.

--------------------

For historical context, this PR was initially a draft fix for #29777 that removed the condition that `this.tests.length > 0`. This turned out to break many tests (https://github.com/web-platform-tests/wpt/issues/29777#issuecomment-896439309), and this PR was repurposed to solve this in a different way.

andreubotella commented:
Closing because the amount of test breakage this change introduces (see https://github.com/web-platform-tests/wpt/issues/29777#issuecomment-896439309) means it's probably best to try other options to solve #29777.

andreubotella commented:
Reopening because replacing `this.tests.length > 0` with `(this.tests.length > 0 || this.pending_remotes.length > 0)` might just do the trick.

andreubotella commented:
PR rebased, squashed, renamed and marked as ready for review.

andreubotella commented:
I've checked the diffs of this PR with the corresponding commit in master (7df27fb) for Chrome, and every difference seems to be due to a fluke that also affects other recent runs. The one exception is `storage-access-api/hasStorageAccess.sub.window.html`, but while debugging that I found bugs with the test itself (see #30320), and after fixing those bugs, the results before and after this change are identical (both for `wpt serve` and `wpt run`). So this PR by itself doesn't seem to be causing any unexpected failures or successes.

jgraham commented on resources/testharness.js:
```diff
@@ -3168,18 +3169,33 @@
     };
 
-    Tests.prototype.fetch_tests_from_worker = function(worker) {
+    Tests.prototype.fetch_tests_from_worker = function(promiseOrWorker) {
```
So I this this is a good change, but technically this kind of API change should have an RFC. Maybe if others are happy with this change we could skip that in this case, but I'd like feedback from at least some of @Ms2ger, @foolip, and @gsnedders 

It definitely requires a docs update in `docs/writing_tests/testharness-api.md`.

andreubotella replied:
I was wondering whether to update the "Determining when all tests are complete" section, but since that section seems to skip a number of the conditions in `Tests.prototype.all_done`, it seems like this change wouldn't be significant enough to mention there.

Ms2ger replied:
I don't have enough of this code paged in to be confident in the change, but I won't object to it.

andreubotella replied:
This is now moot; see https://github.com/web-platform-tests/wpt/pull/29970#issuecomment-916781371

jgraham reviewed (changes requested):
No substantive concerns, but I think we need a docs update, and either an RFC or some agreement that this is a small enough change to skip an RFC.

andreubotella commented:
I just realized that the change to make `fetch_tests_from_worker` take a promise isn't really needed anymore. I added it back when this PR removed the `this.tests.length > 0` condition, since it made SW tests be marked as done before the registration promise resolved. But when I changed the condition to `(this.tests.length > 0 || this.pending_remotes.length > 0)`, now a test file that only fetches tests from a SW wouldn't complete before the remote context is set, so there's no longer any reason to add a dummy pending remote while the promise is pending.

Commit:
Change `Tests.prototype.all_done` to require either tests or pending remotes

`Tests.prototype.all_done` used to treat the presence of at least one
test as a necessary condition for all tests to be done. This would be
false in cases where an exception is thrown inside the `setup` function,
but the code path that runs in that case doesn't use the `all_done()`
function.

However, if an exception is thrown inside the `setup` function in a
worker test, `tests.all_done()` is called in that code path, which
results in the completion message from the worker being ignored. This
change fixes this by changing the condition to require the presence of
either tests or remote contexts.

Closes #29777.

andreubotella commented:
Checked the diffs against master, and there are no regressions that can't be explained as flaky tests.

jgraham reviewed (approved):
Thanks for all the work on this. I know it's hard to get even a one line change to land :)
